Creates stream pool

mruberry opened the pull request:
This PR creates a stream pool per issue #9646. When a new stream is requested, that device it's requested on lazily creates two pools, one low priority and one high priority, of 32 streams each. Streams are returned from these pools round-robin. That is, stream 0 is returned, then stream 1... then stream 31, then stream 0... This PR also takes the opportunity to clean up the stream API, reducing its complexity and verbosity. 

Change notes:

- There are now 3 sets of streams per device, the default stream, the low priority streams, and the high priority streams. These streams live in lazily initialized pools and are destroyed on shutdown.
- All stream refcounting has been removed (the pools pattern replaces it). 
- Setting a stream now sets it on its device. Streams are associated with a device and the previous
requirement to specify that device was unnecessary.
- There is no exposure for setting the flags on a stream. This may also seem like a regression but the flag was always set to cudaStreamNonBlocking.
- Streams are now low or high priority whereas previously the priority could be set with an integer. In practice, however, the range for priorities is -1 to 0 on the latest hardware. -1 is high priority, 0 is low priority (aka default priority). Low vs. high actually clarifies this behavior if people were trying finer separations. (E.g., if someone tried streams with priorities 0, 1, and 2, they would actually all have priority 0, historically, and the intended behavior would not be respected.) 
- Unused THCStream and THCState stream-related functions were removed. 
- A new test of pooling behavior was added in stream_test. 

fyi: @colesbury, @apaszke, @goldsborough

mruberry committed:
Creates stream pool

mruberry committed:
Guards cudaDeviceGetStreamPriorityRange for HIP

mruberry committed:
Guards on device and adds multi-gpu test (to be expanded with CUDAEvent)

mruberry committed:
Fixes narrowing conversion

goldsborough commented on aten/src/ATen/cuda/CUDAStream.cpp:
```diff
@@ -1,57 +1,127 @@
-  static CUDAStreamInternals* default_streams;
+  static constexpr int STREAMS_PER_POOL = 32;
+  static constexpr unsigned int DEFAULT_FLAGS = cudaStreamNonBlocking;
+  static int HIGH_PRIORITY = 0;
```
btw if you're already touching this, our convention in ATen/PyTorch for global constants is `kCamelCaseConstant`, might want to change this

mruberry replied:
Sounds good.

ezyang replied:
Hmm... can we initialize these constants to something that is more obviously uninitialized? It's also a bit annoying that they look like constants, but they are not valid until `initDeviceStreamState` is called... this should be documented here, at least.

ezyang replied:
Since these constants are not initialized for HIP, I'd quite like it if they were ifdef'ed out, so people don't accidentally use them (in their uninitialized form).

yf225 commented:
@colesbury @apaszke Any reviews?

yf225 commented:
@mruberry we probably need a rebase for this PR

@colesbury, @apaszke, @goldsborough any suggestions?

mruberry commented:
Happy to rebase but we should get a review first.

apaszke commented on aten/src/ATen/cuda/CUDAStream.cpp:
```diff
@@ -1,57 +1,127 @@
+    low_priority_streams[device].resize(STREAMS_PER_POOL);
+    high_priority_streams[device].resize(STREAMS_PER_POOL);
+    
+    for (auto i = decltype(STREAMS_PER_POOL){0}; i < STREAMS_PER_POOL; ++i) {
```
This loop is unrolled unnecessarily. I'm pretty sure you could do sth like
```cpp
for (auto streams : {low_priority_streams, high_priority_streams})
```

apaszke commented on aten/src/ATen/cuda/CUDAStream.cpp:
```diff
@@ -1,57 +1,127 @@
+  // Non-default streams
+  static std::deque<std::once_flag> device_flags;
+  static std::deque<std::atomic<int>> low_priority_counters;
+  static std::deque<std::atomic<int>> high_priority_counters;
```
Why do we use a `deque` here?

ezyang replied:
Probably because we're storing `std::atomic<int>` in it, which is not copyable.

mruberry replied:
Ed is, as usual, correct. Not copyable or movable means deque is your default. You can't store these types in a vector.

ezyang replied:
It might still be possible to use a vector here, but we'd have to arrange for the constructor of `high_priority_counters` to occur at a point when we know what the number of GPUs are (because you have to construct the vector with the final number of elements.)

apaszke replied:
You can store those types in a vector, as long as you don't expand it at runtime. I've used vectors of mutexes in the past and they worked fine.

apaszke commented on aten/src/ATen/cuda/CUDAStream.cpp:
```diff
@@ -67,77 +137,64 @@ namespace detail {
+    int modded = raw_idx % STREAMS_PER_POOL;
+    if (raw_idx >= STREAMS_PER_POOL && modded == 0) {
+      counter -= STREAMS_PER_POOL;
+    }
```
This is confusing. You're claiming to generate numbers in the `[0, STREAMS_PER_POOL]` interval, but it really looks like you're doing `[1, STREAMS_PER_POOL]`. Apart from that it seems simpler to just do `return (raw_idx % STREAMS_PER_POOL) + 1`

ezyang replied:
Yes, use an unsigned integer type so that it has well-defined wraparound overflow, and just don't bother trying to force it back in the range. Because `STREAMS_PER_POOL` is a power of two, you won't even have imbalance when you get to the wraparound point. (Actually, it looks like signed integers in std::atomic are defined to uses two's complement, so overflow there is defined too.)

apaszke commented on aten/src/ATen/cuda/CUDAStream.cpp:
```diff
@@ -67,77 +137,64 @@ namespace detail {
-    current_streams[device] = ptr;
-  }
+    const auto idx = get_idx(low_priority_counters[device]);
+    return &low_priority_streams[device][idx];
```
Could we simply assign `low_priority_streams` or `high_priority_streams` to a reference? There's really no reason to duplicate the code.

apaszke reviewed (commented):
(Not a complete review. Some notes)

ezyang commented on aten/src/ATen/cuda/CUDAStream.cpp:
```diff
@@ -1,57 +1,127 @@
+
+  ~CUDAStreamInternals() {
+    if (stream) cudaStreamDestroy(stream);
+  }
```
Doesn't have to be this PR, but it would be nice if `CUDAStreamInternals` stored `cudaStream_t` as a `std::unique_ptr<CUStream_st, CUDAStreamDeleter>` where CUDAStreamDeleter is a functor that calls `cudaStreamDestroy`. Example of this in the cudnn descriptor code.

ezyang commented on aten/src/ATen/cuda/CUDAStream.cpp:
```diff
@@ -1,57 +1,127 @@
+
+  // Non-default streams
+  static std::deque<std::once_flag> device_flags;
+  static std::deque<std::atomic<int>> low_priority_counters;
```
A short comment saying what the counters count would be appreciated. It's a mapping of device to number of in use streams for that device, if I understand correctly.

ezyang commented on aten/src/ATen/cuda/CUDAStream.cpp:
```diff
@@ -1,57 +1,127 @@
+  static std::deque<std::once_flag> device_flags;
+  static std::deque<std::atomic<int>> low_priority_counters;
+  static std::deque<std::atomic<int>> high_priority_counters;
+  static std::vector<std::vector<CUDAStreamInternals>> low_priority_streams;
```
The inner vector is fixed size right? So we can use a `std::array` here.

ezyang commented on aten/src/ATen/cuda/CUDAStream.cpp:
```diff
@@ -1,57 +1,127 @@
       default_streams[i].device = i;
-      default_streams[i].stream = DEFAULT_STREAM;
+      low_priority_counters[i] = 0;
+      high_priority_counters[i] = 0;
```
Pro-tip: you can initialize the counters to zero upon resize using `low_priority_counters.resize(num_gpus, 0)`

ezyang commented on aten/src/ATen/cuda/CUDAStream.cpp:
```diff
@@ -1,57 +1,127 @@
+        , DEFAULT_FLAGS
+        , HIGH_PRIORITY));
+      #else 
+        AT_CUDA_CHECK(cudaStreamCreateWithFlags(
```
Maybe a comment here with something like, "HIP doesn't support stream priority, so we just create two streams with the same priority."

ezyang reviewed (dismissed):
At a high level, it all looks good. All my comments are just lower level nits. In terms of priority, the most important change to make for me is changing how counter wraparound works.

We might need some documentation about how the streams used here should be short lived. The discussion in the upstream issue was nicely detailed, but people are not likely to see it once this merges.

mruberry commented:
Thanks for taking a look @ezyang, @apaszke. Suggestions look good. Really like your point about commenting, @ezyang. I'll get us an update soon (have to finish splitting the fusion compiler first).

mruberry committed:
merges and updates per comments

mruberry commented:
I merged with master and made the following changes:

- Added a note to CUDAStream.h and additional comments to CUDAStream.cpp, clarifying the use of counters and flags, in particular.
- Updated constant names per @goldsborough, guarded values per @ezyang, used arrays per @ezyang, also simplified high vs low priority so these values are simply initialized properly (the prior approach was needlessly general)
- Changed the atomic counters to uint32_t, simplified the round-robin logic by allowing overflow

I did not:
- Update CUDAStreamInternals to use a unique ptr with a custom deleter, which I agree is probably more elegant but also not necessary right now.
- Change the loop unroll since I think it's good enough
- Merge the high priority and low priority acquisitions in CUDAStream_createStream(). The lines of code here can be reduced but I think the current statement is clear and the logic duplication is very small.

mruberry committed:
improves comment
